src/Data/Tree/AVL/Internal.hs

```diff
@@ -51,148 +64,177 @@ data Tilt
-    minBound)
-
-setLeft    :: Setter' (Map h k v) (Map h k v)
-setRight   :: Setter' (Map h k v) (Map h k v)
-setValue   :: Setter' (Map h k v) v
-setNextKey :: Setter' (Map h k v) k
-setPrevKey :: Setter' (Map h k v) k
-
-setLeft    = _Fix.mlLeft
-setRight   = _Fix.mlRight
-setValue   = _Fix.mlValue
-setNextKey = _Fix.mlNextKey
-setPrevKey = _Fix.mlPrevKey
-
-branching :: Getter (Map h k v) (Maybe (Branching h k v))
-branching = to $ \case
-  Branch _ _ _ _ t l r -> Just (Branching l r t)
-  _                    -> Nothing
-
-terminal :: Getter (Map h k v) (Maybe (Terminal h k v))
-terminal = to $ \case
-  Leaf _ _ k v n p -> Just (Terminal k v n p)
-  _                -> Nothing
-
-vacuous :: Getter (Map h k v) (Maybe Vacuous)
-vacuous = to $ \case
-  Empty _ _ -> Just (Vacuous)
-  _         -> Nothing
 
--- | Recalculate 'rootHash' of the node.
---   Does nothing on 'Pruned' node.
-rehash :: Hash h k v => Map h k v -> Map h k v
-rehash node = case tree of
-  MLPruned {} -> node
-  _           -> Fix $ tree & mlHash .~ hashOf cleaned
-  where
-    cleaned  = fmap (^.rootHash) $ tree & mlHash .~ def
-    Fix tree = node
+-- | Class, representing DB layer, capable of storing 'isolate'd nodes.
+--   The 'store' is an idempotent operation.
+class (MonadCatch m, MonadIO m) => KVStoreMonad h m where
+    retrieve :: Binary v => h -> m v
+    store    :: Binary v => h -> v -> m ()
+
+-- | Exception to be thrown when node with given hashkey is missing.
+data NotFound k = NotFound k
+    deriving (Show, Typeable)
+
+instance (Show k, Typeable k) => Exception (NotFound k)
+
+-- | Debug preview of the tree.
+--showMap :: (Show h, Show k, Show v) => Map h k v -> String
+--showMap = drawTree . asTree
+--  where
+--    asTree = \case
+--      Free (MLBranch _ _mk _ck t  l r) -> Tree.Node ("-< "  ++ show (t)) [asTree r, asTree l]
+--      Free (MLLeaf   _  k  _v _n _p)   -> Tree.Node ("<3- " ++ show (k)) []
+--      Free (MLEmpty  _)                -> Tree.Node ("--")               []
+--      Pure  h                          -> Tree.Node ("Ref " ++ show h)   []
+
+instance (Show h, Show k, Show v, Show self) => Show (MapLayer h k v self) where
+    show = \case
+      MLBranch h _mk _ck  t  l r -> "Branch" ++ show (h, t, l, r)
+      MLLeaf   h  k  _v  _n _p   -> "Leaf"   ++ show (h, k)
+      MLEmpty  h                 -> "--"     ++ show (h)
 
 -- | Interface for calculating hash of the 'Map' node.
 class
-    (Ord k, Show k, Show h, Show v, Bounded k, Eq h, Default h)
+    (Ord k, Show k, Show h, Hashable h, Typeable h, Show v, Bounded k, Eq h, Default h)
       =>
     Hash h k v
   where
     hashOf :: MapLayer h k v h -> h
 
+-- | Umbrella class to grab all the required capabilities for tree to operate (and be debugged!).
+class    (Ord h, Typeable k, Hash h k v, Binary h, Binary k, Binary v, KVStoreMonad h m) => Stores h k v m where
+instance (Ord h, Typeable k, Hash h k v, Binary h, Binary k, Binary v, KVStoreMonad h m) => Stores h k v m where
```

Comment:
When it is a synonym, on actual error you get message that "there is a no KVStoreMonad m instance" at the point of "Stores h k v m", which is misleading. I actually tried it and was cofused by the errors.

The problem arises because Ghc blindly replaces all type synonyms with their value.

That's why i generally dislike type synonyms